// @datocms/cma-client/dist/cjs/generated/resources/AuditLogEvent.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Utils = __importStar(require("@datocms/rest-client-utils"));
const BaseResource_1 = __importDefault(require("../../BaseResource"));
class AuditLogEvent extends BaseResource_1.default {
    /**
     * List Audit Log events
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/audit-log_event/query
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    query(body) {
        return this.rawQuery(Utils.serializeRequestBody(body, {
            type: 'audit_log_query',
            attributes: ['filter', 'next_token', 'detailed_log'],
            relationships: [],
        })).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * List Audit Log events
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/audit-log_event/query
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawQuery(body) {
        return this.client.request({
            method: 'POST',
            url: '/audit-log-events/query',
            body,
        });
    }
}
exports.default = AuditLogEvent;
AuditLogEvent.TYPE = 'audit_log_event';
