// @datocms/cma-client/dist/cjs/generated/resources/Fieldset.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Utils = __importStar(require("@datocms/rest-client-utils"));
const BaseResource_1 = __importDefault(require("../../BaseResource"));
class Fieldset extends BaseResource_1.default {
    /**
     * Create a new fieldset
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/fieldset/create
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    create(itemTypeId, body) {
        return this.rawCreate(Utils.toId(itemTypeId), Utils.serializeRequestBody(body, {
            type: 'fieldset',
            attributes: [
                'title',
                'hint',
                'position',
                'collapsible',
                'start_collapsed',
            ],
            relationships: [],
        })).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Create a new fieldset
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/fieldset/create
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawCreate(itemTypeId, body) {
        return this.client.request({
            method: 'POST',
            url: `/item-types/${itemTypeId}/fieldsets`,
            body,
        });
    }
    /**
     * Update a fieldset
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/fieldset/update
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    update(fieldsetId, body) {
        return this.rawUpdate(Utils.toId(fieldsetId), Utils.serializeRequestBody(body, {
            id: Utils.toId(fieldsetId),
            type: 'fieldset',
            attributes: [
                'title',
                'hint',
                'position',
                'collapsible',
                'start_collapsed',
            ],
            relationships: [],
        })).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Update a fieldset
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/fieldset/update
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawUpdate(fieldsetId, body) {
        return this.client.request({
            method: 'PUT',
            url: `/fieldsets/${fieldsetId}`,
            body,
        });
    }
    /**
     * List all fieldsets of a model/block
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/fieldset/instances
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    list(itemTypeId) {
        return this.rawList(Utils.toId(itemTypeId)).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * List all fieldsets of a model/block
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/fieldset/instances
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawList(itemTypeId) {
        return this.client.request({
            method: 'GET',
            url: `/item-types/${itemTypeId}/fieldsets`,
        });
    }
    /**
     * Retrieve a fieldset
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/fieldset/self
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    find(fieldsetId) {
        return this.rawFind(Utils.toId(fieldsetId)).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Retrieve a fieldset
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/fieldset/self
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawFind(fieldsetId) {
        return this.client.request({
            method: 'GET',
            url: `/fieldsets/${fieldsetId}`,
        });
    }
    /**
     * Delete a fieldset
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/fieldset/destroy
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    destroy(fieldsetId) {
        return this.rawDestroy(Utils.toId(fieldsetId)).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Delete a fieldset
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/fieldset/destroy
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawDestroy(fieldsetId) {
        return this.client.request({
            method: 'DELETE',
            url: `/fieldsets/${fieldsetId}`,
        });
    }
}
exports.default = Fieldset;
Fieldset.TYPE = 'fieldset';
