// @datocms/cma-client/dist/cjs/generated/resources/Field.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var Utils = __importStar(require("@datocms/rest-client-utils"));
var BaseResource_1 = __importDefault(require("../../BaseResource"));
var Field = /** @class */ (function (_super) {
    __extends(Field, _super);
    function Field() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Create a new field
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/field/create
     */
    Field.prototype.create = function (itemTypeId, body) {
        return this.rawCreate(Utils.toId(itemTypeId), Utils.serializeRequestBody(body, {
            type: 'field',
            attributes: [
                'label',
                'field_type',
                'api_key',
                'localized',
                'validators',
                'appeareance',
                'appearance',
                'position',
                'hint',
                'default_value',
            ],
            relationships: ['fieldset'],
        })).then(function (body) {
            return Utils.deserializeResponseBody(body);
        });
    };
    /**
     * Create a new field
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/field/create
     */
    Field.prototype.rawCreate = function (itemTypeId, body) {
        return this.client.request({
            method: 'POST',
            url: "/item-types/".concat(itemTypeId, "/fields"),
            body: body,
        });
    };
    /**
     * Update a field
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/field/update
     */
    Field.prototype.update = function (fieldId, body) {
        return this.rawUpdate(Utils.toId(fieldId), Utils.serializeRequestBody(body, {
            id: Utils.toId(fieldId),
            type: 'field',
            attributes: [
                'default_value',
                'label',
                'api_key',
                'localized',
                'validators',
                'appeareance',
                'appearance',
                'position',
                'field_type',
                'hint',
            ],
            relationships: ['fieldset'],
        })).then(function (body) {
            return Utils.deserializeResponseBody(body);
        });
    };
    /**
     * Update a field
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/field/update
     */
    Field.prototype.rawUpdate = function (fieldId, body) {
        return this.client.request({
            method: 'PUT',
            url: "/fields/".concat(fieldId),
            body: body,
        });
    };
    /**
     * List all fields
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/field/instances
     */
    Field.prototype.list = function (itemTypeId) {
        return this.rawList(Utils.toId(itemTypeId)).then(function (body) {
            return Utils.deserializeResponseBody(body);
        });
    };
    /**
     * List all fields
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/field/instances
     */
    Field.prototype.rawList = function (itemTypeId) {
        return this.client.request({
            method: 'GET',
            url: "/item-types/".concat(itemTypeId, "/fields"),
        });
    };
    /**
     * List referenced fields
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/field/referencing
     *
     * @deprecated This API call is to be considered private and might change without notice
     */
    Field.prototype.referencing = function (itemTypeId) {
        return this.rawReferencing(Utils.toId(itemTypeId)).then(function (body) {
            return Utils.deserializeResponseBody(body);
        });
    };
    /**
     * List referenced fields
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/field/referencing
     *
     * @deprecated This API call is to be considered private and might change without notice
     */
    Field.prototype.rawReferencing = function (itemTypeId) {
        return this.client.request({
            method: 'GET',
            url: "/item-types/".concat(itemTypeId, "/fields/referencing"),
        });
    };
    /**
     * List related fields
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/field/related
     *
     * @deprecated This API call is to be considered private and might change without notice
     */
    Field.prototype.related = function (itemTypeId) {
        return this.rawRelated(Utils.toId(itemTypeId)).then(function (body) {
            return Utils.deserializeResponseBody(body);
        });
    };
    /**
     * List related fields
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/field/related
     *
     * @deprecated This API call is to be considered private and might change without notice
     */
    Field.prototype.rawRelated = function (itemTypeId) {
        return this.client.request({
            method: 'GET',
            url: "/item-types/".concat(itemTypeId, "/fields/related"),
        });
    };
    /**
     * Retrieve a field
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/field/self
     */
    Field.prototype.find = function (fieldId) {
        return this.rawFind(Utils.toId(fieldId)).then(function (body) {
            return Utils.deserializeResponseBody(body);
        });
    };
    /**
     * Retrieve a field
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/field/self
     */
    Field.prototype.rawFind = function (fieldId) {
        return this.client.request({
            method: 'GET',
            url: "/fields/".concat(fieldId),
        });
    };
    /**
     * Delete a field
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/field/destroy
     */
    Field.prototype.destroy = function (fieldId) {
        return this.rawDestroy(Utils.toId(fieldId)).then(function (body) {
            return Utils.deserializeResponseBody(body);
        });
    };
    /**
     * Delete a field
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/field/destroy
     */
    Field.prototype.rawDestroy = function (fieldId) {
        return this.client.request({
            method: 'DELETE',
            url: "/fields/".concat(fieldId),
        });
    };
    /**
     * Duplicate a field
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/field/duplicate
     */
    Field.prototype.duplicate = function (fieldId) {
        return this.rawDuplicate(Utils.toId(fieldId)).then(function (body) {
            return Utils.deserializeResponseBody(body);
        });
    };
    /**
     * Duplicate a field
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/field/duplicate
     */
    Field.prototype.rawDuplicate = function (fieldId) {
        return this.client.request({
            method: 'POST',
            url: "/fields/".concat(fieldId, "/duplicate"),
        });
    };
    Field.TYPE = 'field';
    return Field;
}(BaseResource_1.default));
exports.default = Field;
