// @datocms/cma-client/dist/cjs/generated/resources/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Workflow = exports.Site = exports.UploadSmartTag = exports.UploadTag = exports.UsageCounter = exports.DailyUsage = exports.PublicInfo = exports.WhiteLabelSettings = exports.SsoSettings = exports.SsoGroup = exports.EditingSession = exports.SiteInvitation = exports.UploadFilter = exports.ItemTypeFilter = exports.BuildTrigger = exports.WebhookCall = exports.Webhook = exports.MaintenanceMode = exports.Environment = exports.SearchResult = exports.ScheduledUnpublishing = exports.ScheduledPublication = exports.UploadRequest = exports.Upload = exports.ItemVersion = exports.Item = exports.BuildEvent = exports.SubscriptionFeature = exports.SubscriptionLimit = exports.JobResult = exports.Plugin = exports.AccessToken = exports.Session = exports.Fieldset = exports.Field = exports.ItemType = exports.MenuItem = exports.AuditLogEvent = exports.SsoUser = exports.User = exports.Role = void 0;
var Role_1 = require("./Role");
Object.defineProperty(exports, "Role", { enumerable: true, get: function () { return __importDefault(Role_1).default; } });
var User_1 = require("./User");
Object.defineProperty(exports, "User", { enumerable: true, get: function () { return __importDefault(User_1).default; } });
var SsoUser_1 = require("./SsoUser");
Object.defineProperty(exports, "SsoUser", { enumerable: true, get: function () { return __importDefault(SsoUser_1).default; } });
var AuditLogEvent_1 = require("./AuditLogEvent");
Object.defineProperty(exports, "AuditLogEvent", { enumerable: true, get: function () { return __importDefault(AuditLogEvent_1).default; } });
var MenuItem_1 = require("./MenuItem");
Object.defineProperty(exports, "MenuItem", { enumerable: true, get: function () { return __importDefault(MenuItem_1).default; } });
var ItemType_1 = require("./ItemType");
Object.defineProperty(exports, "ItemType", { enumerable: true, get: function () { return __importDefault(ItemType_1).default; } });
var Field_1 = require("./Field");
Object.defineProperty(exports, "Field", { enumerable: true, get: function () { return __importDefault(Field_1).default; } });
var Fieldset_1 = require("./Fieldset");
Object.defineProperty(exports, "Fieldset", { enumerable: true, get: function () { return __importDefault(Fieldset_1).default; } });
var Session_1 = require("./Session");
Object.defineProperty(exports, "Session", { enumerable: true, get: function () { return __importDefault(Session_1).default; } });
var AccessToken_1 = require("./AccessToken");
Object.defineProperty(exports, "AccessToken", { enumerable: true, get: function () { return __importDefault(AccessToken_1).default; } });
var Plugin_1 = require("./Plugin");
Object.defineProperty(exports, "Plugin", { enumerable: true, get: function () { return __importDefault(Plugin_1).default; } });
var JobResult_1 = require("./JobResult");
Object.defineProperty(exports, "JobResult", { enumerable: true, get: function () { return __importDefault(JobResult_1).default; } });
var SubscriptionLimit_1 = require("./SubscriptionLimit");
Object.defineProperty(exports, "SubscriptionLimit", { enumerable: true, get: function () { return __importDefault(SubscriptionLimit_1).default; } });
var SubscriptionFeature_1 = require("./SubscriptionFeature");
Object.defineProperty(exports, "SubscriptionFeature", { enumerable: true, get: function () { return __importDefault(SubscriptionFeature_1).default; } });
var BuildEvent_1 = require("./BuildEvent");
Object.defineProperty(exports, "BuildEvent", { enumerable: true, get: function () { return __importDefault(BuildEvent_1).default; } });
var Item_1 = require("./Item");
Object.defineProperty(exports, "Item", { enumerable: true, get: function () { return __importDefault(Item_1).default; } });
var ItemVersion_1 = require("./ItemVersion");
Object.defineProperty(exports, "ItemVersion", { enumerable: true, get: function () { return __importDefault(ItemVersion_1).default; } });
var Upload_1 = require("./Upload");
Object.defineProperty(exports, "Upload", { enumerable: true, get: function () { return __importDefault(Upload_1).default; } });
var UploadRequest_1 = require("./UploadRequest");
Object.defineProperty(exports, "UploadRequest", { enumerable: true, get: function () { return __importDefault(UploadRequest_1).default; } });
var ScheduledPublication_1 = require("./ScheduledPublication");
Object.defineProperty(exports, "ScheduledPublication", { enumerable: true, get: function () { return __importDefault(ScheduledPublication_1).default; } });
var ScheduledUnpublishing_1 = require("./ScheduledUnpublishing");
Object.defineProperty(exports, "ScheduledUnpublishing", { enumerable: true, get: function () { return __importDefault(ScheduledUnpublishing_1).default; } });
var SearchResult_1 = require("./SearchResult");
Object.defineProperty(exports, "SearchResult", { enumerable: true, get: function () { return __importDefault(SearchResult_1).default; } });
var Environment_1 = require("./Environment");
Object.defineProperty(exports, "Environment", { enumerable: true, get: function () { return __importDefault(Environment_1).default; } });
var MaintenanceMode_1 = require("./MaintenanceMode");
Object.defineProperty(exports, "MaintenanceMode", { enumerable: true, get: function () { return __importDefault(MaintenanceMode_1).default; } });
var Webhook_1 = require("./Webhook");
Object.defineProperty(exports, "Webhook", { enumerable: true, get: function () { return __importDefault(Webhook_1).default; } });
var WebhookCall_1 = require("./WebhookCall");
Object.defineProperty(exports, "WebhookCall", { enumerable: true, get: function () { return __importDefault(WebhookCall_1).default; } });
var BuildTrigger_1 = require("./BuildTrigger");
Object.defineProperty(exports, "BuildTrigger", { enumerable: true, get: function () { return __importDefault(BuildTrigger_1).default; } });
var ItemTypeFilter_1 = require("./ItemTypeFilter");
Object.defineProperty(exports, "ItemTypeFilter", { enumerable: true, get: function () { return __importDefault(ItemTypeFilter_1).default; } });
var UploadFilter_1 = require("./UploadFilter");
Object.defineProperty(exports, "UploadFilter", { enumerable: true, get: function () { return __importDefault(UploadFilter_1).default; } });
var SiteInvitation_1 = require("./SiteInvitation");
Object.defineProperty(exports, "SiteInvitation", { enumerable: true, get: function () { return __importDefault(SiteInvitation_1).default; } });
var EditingSession_1 = require("./EditingSession");
Object.defineProperty(exports, "EditingSession", { enumerable: true, get: function () { return __importDefault(EditingSession_1).default; } });
var SsoGroup_1 = require("./SsoGroup");
Object.defineProperty(exports, "SsoGroup", { enumerable: true, get: function () { return __importDefault(SsoGroup_1).default; } });
var SsoSettings_1 = require("./SsoSettings");
Object.defineProperty(exports, "SsoSettings", { enumerable: true, get: function () { return __importDefault(SsoSettings_1).default; } });
var WhiteLabelSettings_1 = require("./WhiteLabelSettings");
Object.defineProperty(exports, "WhiteLabelSettings", { enumerable: true, get: function () { return __importDefault(WhiteLabelSettings_1).default; } });
var PublicInfo_1 = require("./PublicInfo");
Object.defineProperty(exports, "PublicInfo", { enumerable: true, get: function () { return __importDefault(PublicInfo_1).default; } });
var DailyUsage_1 = require("./DailyUsage");
Object.defineProperty(exports, "DailyUsage", { enumerable: true, get: function () { return __importDefault(DailyUsage_1).default; } });
var UsageCounter_1 = require("./UsageCounter");
Object.defineProperty(exports, "UsageCounter", { enumerable: true, get: function () { return __importDefault(UsageCounter_1).default; } });
var UploadTag_1 = require("./UploadTag");
Object.defineProperty(exports, "UploadTag", { enumerable: true, get: function () { return __importDefault(UploadTag_1).default; } });
var UploadSmartTag_1 = require("./UploadSmartTag");
Object.defineProperty(exports, "UploadSmartTag", { enumerable: true, get: function () { return __importDefault(UploadSmartTag_1).default; } });
var Site_1 = require("./Site");
Object.defineProperty(exports, "Site", { enumerable: true, get: function () { return __importDefault(Site_1).default; } });
var Workflow_1 = require("./Workflow");
Object.defineProperty(exports, "Workflow", { enumerable: true, get: function () { return __importDefault(Workflow_1).default; } });
