// @datocms/cma-client/dist/cjs/generated/resources/ItemTypeFilter.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var Utils = __importStar(require("@datocms/rest-client-utils"));
var BaseResource_1 = __importDefault(require("../../BaseResource"));
var ItemTypeFilter = /** @class */ (function (_super) {
    __extends(ItemTypeFilter, _super);
    function ItemTypeFilter() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Create a new filter
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/item-type_filter/create
     */
    ItemTypeFilter.prototype.create = function (body) {
        return this.rawCreate(Utils.serializeRequestBody(body, {
            type: 'item_type_filter',
            attributes: ['name', 'filter', 'shared'],
            relationships: ['item_type'],
        })).then(function (body) {
            return Utils.deserializeResponseBody(body);
        });
    };
    /**
     * Create a new filter
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/item-type_filter/create
     */
    ItemTypeFilter.prototype.rawCreate = function (body) {
        return this.client.request({
            method: 'POST',
            url: "/item-type-filters",
            body: body,
        });
    };
    /**
     * Update a filter
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/item-type_filter/update
     */
    ItemTypeFilter.prototype.update = function (itemTypeFilterId, body) {
        return this.rawUpdate(Utils.toId(itemTypeFilterId), Utils.serializeRequestBody(body, {
            id: Utils.toId(itemTypeFilterId),
            type: 'item_type_filter',
            attributes: ['name', 'shared', 'filter'],
            relationships: [],
        })).then(function (body) {
            return Utils.deserializeResponseBody(body);
        });
    };
    /**
     * Update a filter
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/item-type_filter/update
     */
    ItemTypeFilter.prototype.rawUpdate = function (itemTypeFilterId, body) {
        return this.client.request({
            method: 'PUT',
            url: "/item-type-filters/".concat(itemTypeFilterId),
            body: body,
        });
    };
    /**
     * List all filters
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/item-type_filter/instances
     */
    ItemTypeFilter.prototype.list = function () {
        return this.rawList().then(function (body) {
            return Utils.deserializeResponseBody(body);
        });
    };
    /**
     * List all filters
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/item-type_filter/instances
     */
    ItemTypeFilter.prototype.rawList = function () {
        return this.client.request({
            method: 'GET',
            url: "/item-type-filters",
        });
    };
    /**
     * Retrieve a filter
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/item-type_filter/self
     */
    ItemTypeFilter.prototype.find = function (itemTypeFilterId) {
        return this.rawFind(Utils.toId(itemTypeFilterId)).then(function (body) {
            return Utils.deserializeResponseBody(body);
        });
    };
    /**
     * Retrieve a filter
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/item-type_filter/self
     */
    ItemTypeFilter.prototype.rawFind = function (itemTypeFilterId) {
        return this.client.request({
            method: 'GET',
            url: "/item-type-filters/".concat(itemTypeFilterId),
        });
    };
    /**
     * Delete a filter
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/item-type_filter/destroy
     */
    ItemTypeFilter.prototype.destroy = function (itemTypeFilterId) {
        return this.rawDestroy(Utils.toId(itemTypeFilterId)).then(function (body) {
            return Utils.deserializeResponseBody(body);
        });
    };
    /**
     * Delete a filter
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/item-type_filter/destroy
     */
    ItemTypeFilter.prototype.rawDestroy = function (itemTypeFilterId) {
        return this.client.request({
            method: 'DELETE',
            url: "/item-type-filters/".concat(itemTypeFilterId),
        });
    };
    ItemTypeFilter.TYPE = 'item_type_filter';
    return ItemTypeFilter;
}(BaseResource_1.default));
exports.default = ItemTypeFilter;
