// @datocms/cma-client/dist/cjs/generated/resources/ItemTypeFilter.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Utils = __importStar(require("@datocms/rest-client-utils"));
const BaseResource_1 = __importDefault(require("../../BaseResource"));
class ItemTypeFilter extends BaseResource_1.default {
    /**
     * Create a new filter
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/item-type_filter/create
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    create(body) {
        return this.rawCreate(Utils.serializeRequestBody(body, {
            type: 'item_type_filter',
            attributes: ['name', 'filter', 'columns', 'order_by', 'shared'],
            relationships: ['item_type'],
        })).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Create a new filter
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/item-type_filter/create
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawCreate(body) {
        return this.client.request({
            method: 'POST',
            url: '/item-type-filters',
            body,
        });
    }
    /**
     * Update a filter
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/item-type_filter/update
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    update(itemTypeFilterId, body) {
        return this.rawUpdate(Utils.toId(itemTypeFilterId), Utils.serializeRequestBody(body, {
            id: Utils.toId(itemTypeFilterId),
            type: 'item_type_filter',
            attributes: ['name', 'columns', 'order_by', 'shared', 'filter'],
            relationships: ['item_type'],
        })).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Update a filter
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/item-type_filter/update
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawUpdate(itemTypeFilterId, body) {
        return this.client.request({
            method: 'PUT',
            url: `/item-type-filters/${itemTypeFilterId}`,
            body,
        });
    }
    /**
     * List all filters
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/item-type_filter/instances
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    list() {
        return this.rawList().then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * List all filters
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/item-type_filter/instances
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawList() {
        return this.client.request({
            method: 'GET',
            url: '/item-type-filters',
        });
    }
    /**
     * Retrieve a filter
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/item-type_filter/self
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    find(itemTypeFilterId) {
        return this.rawFind(Utils.toId(itemTypeFilterId)).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Retrieve a filter
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/item-type_filter/self
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawFind(itemTypeFilterId) {
        return this.client.request({
            method: 'GET',
            url: `/item-type-filters/${itemTypeFilterId}`,
        });
    }
    /**
     * Delete a filter
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/item-type_filter/destroy
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    destroy(itemTypeFilterId) {
        return this.rawDestroy(Utils.toId(itemTypeFilterId)).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Delete a filter
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/item-type_filter/destroy
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawDestroy(itemTypeFilterId) {
        return this.client.request({
            method: 'DELETE',
            url: `/item-type-filters/${itemTypeFilterId}`,
        });
    }
}
exports.default = ItemTypeFilter;
ItemTypeFilter.TYPE = 'item_type_filter';
