// @datocms/cma-client/dist/cjs/generated/resources/MaintenanceMode.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Utils = __importStar(require("@datocms/rest-client-utils"));
const BaseResource_1 = __importDefault(require("../../BaseResource"));
class MaintenanceMode extends BaseResource_1.default {
    /**
     * Retrieve maintenence mode
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/maintenance-mode/self
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    find() {
        return this.rawFind().then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Retrieve maintenence mode
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/maintenance-mode/self
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawFind() {
        return this.client.request({
            method: 'GET',
            url: '/maintenance-mode',
        });
    }
    /**
     * Activate maintenance mode: this means that the primary environment will be read-only
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/maintenance-mode/activate
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    activate(queryParams) {
        return this.rawActivate(queryParams).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Activate maintenance mode: this means that the primary environment will be read-only
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/maintenance-mode/activate
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawActivate(queryParams) {
        return this.client.request({
            method: 'PUT',
            url: '/maintenance-mode/activate',
            queryParams,
        });
    }
    /**
     * De-activate maintenance mode
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/maintenance-mode/deactivate
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    deactivate() {
        return this.rawDeactivate().then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * De-activate maintenance mode
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/maintenance-mode/deactivate
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawDeactivate() {
        return this.client.request({
            method: 'PUT',
            url: '/maintenance-mode/deactivate',
        });
    }
}
exports.default = MaintenanceMode;
MaintenanceMode.TYPE = 'maintenance_mode';
