// @datocms/cma-client/dist/cjs/generated/resources/ScheduledPublication.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Utils = __importStar(require("@datocms/rest-client-utils"));
const BaseResource_1 = __importDefault(require("../../BaseResource"));
class ScheduledPublication extends BaseResource_1.default {
    /**
     * Create a new scheduled publication
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/scheduled-publication/create
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    create(itemId, body) {
        return this.rawCreate(Utils.toId(itemId), Utils.serializeRequestBody(body, {
            type: 'scheduled_publication',
            attributes: ['publication_scheduled_at', 'selective_publication'],
            relationships: [],
        })).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Create a new scheduled publication
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/scheduled-publication/create
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawCreate(itemId, body) {
        return this.client.request({
            method: 'POST',
            url: `/items/${itemId}/scheduled-publication`,
            body,
        });
    }
    /**
     * Delete a scheduled publication
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/scheduled-publication/destroy
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    destroy(itemId) {
        return this.rawDestroy(Utils.toId(itemId)).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Delete a scheduled publication
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/scheduled-publication/destroy
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawDestroy(itemId) {
        return this.client
            .request({
            method: 'DELETE',
            url: `/items/${itemId}/scheduled-publication`,
        })
            .then(Utils.deserializeRawResponseBodyWithItems);
    }
}
exports.default = ScheduledPublication;
ScheduledPublication.TYPE = 'scheduled_publication';
