// @datocms/cma-client/dist/cjs/generated/resources/SchemaMenuItem.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Utils = __importStar(require("@datocms/rest-client-utils"));
const BaseResource_1 = __importDefault(require("../../BaseResource"));
class SchemaMenuItem extends BaseResource_1.default {
    /**
     * Create a new schema menu item
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/schema-menu_item/create
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    create(body) {
        return this.rawCreate(Utils.serializeRequestBody(body, {
            type: 'schema_menu_item',
            attributes: ['label', 'position', 'kind'],
            relationships: ['item_type', 'parent'],
        })).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Create a new schema menu item
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/schema-menu_item/create
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawCreate(body) {
        return this.client.request({
            method: 'POST',
            url: '/schema-menu-items',
            body,
        });
    }
    /**
     * Update a schema menu item
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/schema-menu_item/update
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    update(schemaMenuItemId, body) {
        return this.rawUpdate(Utils.toId(schemaMenuItemId), Utils.serializeRequestBody(body, {
            id: Utils.toId(schemaMenuItemId),
            type: 'schema_menu_item',
            attributes: ['label', 'position', 'kind'],
            relationships: ['item_type', 'parent', 'children'],
        })).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Update a schema menu item
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/schema-menu_item/update
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawUpdate(schemaMenuItemId, body) {
        return this.client.request({
            method: 'PUT',
            url: `/schema-menu-items/${schemaMenuItemId}`,
            body,
        });
    }
    /**
     * List all schema menu items
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/schema-menu_item/instances
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    list(queryParams) {
        return this.rawList(queryParams).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * List all schema menu items
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/schema-menu_item/instances
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawList(queryParams) {
        return this.client.request({
            method: 'GET',
            url: '/schema-menu-items',
            queryParams,
        });
    }
    /**
     * Retrieve a schema menu item
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/schema-menu_item/self
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    find(schemaMenuItemId) {
        return this.rawFind(Utils.toId(schemaMenuItemId)).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Retrieve a schema menu item
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/schema-menu_item/self
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawFind(schemaMenuItemId) {
        return this.client.request({
            method: 'GET',
            url: `/schema-menu-items/${schemaMenuItemId}`,
        });
    }
    /**
     * Delete a schema menu item
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/schema-menu_item/destroy
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    destroy(schemaMenuItemId) {
        return this.rawDestroy(Utils.toId(schemaMenuItemId)).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Delete a schema menu item
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/schema-menu_item/destroy
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawDestroy(schemaMenuItemId) {
        return this.client.request({
            method: 'DELETE',
            url: `/schema-menu-items/${schemaMenuItemId}`,
        });
    }
    /**
     * Reorders a set of schema menu items
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/schema-menu_item/reorder
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     *
     * @deprecated This API call is to be considered private and might change without notice
     */
    reorder(body) {
        return this.rawReorder(Utils.serializeRequestBody(body, {
            type: 'schema_menu_item',
            attributes: ['position'],
            relationships: ['parent'],
        })).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Reorders a set of schema menu items
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/schema-menu_item/reorder
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     *
     * @deprecated This API call is to be considered private and might change without notice
     */
    rawReorder(body) {
        return this.client.request({
            method: 'POST',
            url: '/schema-menu-items/reorder',
            body,
        });
    }
}
exports.default = SchemaMenuItem;
SchemaMenuItem.TYPE = 'schema_menu_item';
