// @datocms/cma-client/dist/cjs/generated/resources/SsoGroup.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Utils = __importStar(require("@datocms/rest-client-utils"));
const BaseResource_1 = __importDefault(require("../../BaseResource"));
class SsoGroup extends BaseResource_1.default {
    /**
     * List all SSO groups
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/sso-group/instances
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    list() {
        return this.rawList().then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * List all SSO groups
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/sso-group/instances
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawList() {
        return this.client.request({
            method: 'GET',
            url: '/sso-groups',
        });
    }
    /**
     * Sync SSO provider groups to DatoCMS roles
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/sso-group/copy_roles
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    copyRoles(ssoGroupId) {
        return this.rawCopyRoles(Utils.toId(ssoGroupId)).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Sync SSO provider groups to DatoCMS roles
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/sso-group/copy_roles
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawCopyRoles(ssoGroupId) {
        return this.client.request({
            method: 'POST',
            url: `/sso-groups/${ssoGroupId}/copy-roles`,
        });
    }
    /**
     * Update a SSO group
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/sso-group/update
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    update(ssoGroupId, body) {
        return this.rawUpdate(Utils.toId(ssoGroupId), Utils.serializeRequestBody(body, {
            id: Utils.toId(ssoGroupId),
            type: 'sso_group',
            attributes: ['priority'],
            relationships: ['role'],
        })).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Update a SSO group
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/sso-group/update
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawUpdate(ssoGroupId, body) {
        return this.client.request({
            method: 'PUT',
            url: `/sso-groups/${ssoGroupId}`,
            body,
        });
    }
    /**
     * Delete a group
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/sso-group/destroy
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    destroy(ssoGroupId) {
        return this.rawDestroy(Utils.toId(ssoGroupId)).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Delete a group
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/sso-group/destroy
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawDestroy(ssoGroupId) {
        return this.client.request({
            method: 'DELETE',
            url: `/sso-groups/${ssoGroupId}`,
        });
    }
}
exports.default = SsoGroup;
SsoGroup.TYPE = 'sso_group';
