// @datocms/cma-client/dist/cjs/generated/resources/SsoSettings.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Utils = __importStar(require("@datocms/rest-client-utils"));
const BaseResource_1 = __importDefault(require("../../BaseResource"));
class SsoSettings extends BaseResource_1.default {
    /**
     * Retrieve SSO Settings
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/sso-settings/self
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    find() {
        return this.rawFind().then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Retrieve SSO Settings
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/sso-settings/self
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawFind() {
        return this.client.request({
            method: 'GET',
            url: '/sso-settings',
        });
    }
    /**
     * Generate SSO token
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/sso-settings/generate_token
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    generateToken() {
        return this.rawGenerateToken().then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Generate SSO token
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/sso-settings/generate_token
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawGenerateToken() {
        return this.client.request({
            method: 'PUT',
            url: '/sso-settings/generate-token',
        });
    }
    /**
     * Update SSO Settings
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/sso-settings/update
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    update(body) {
        return this.rawUpdate(Utils.serializeRequestBody(body, {
            type: 'sso_settings',
            attributes: ['idp_saml_metadata_url', 'idp_saml_metadata_xml'],
            relationships: ['default_role'],
        })).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Update SSO Settings
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/sso-settings/update
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawUpdate(body) {
        return this.client.request({
            method: 'PUT',
            url: '/sso-settings',
            body,
        });
    }
}
exports.default = SsoSettings;
SsoSettings.TYPE = 'sso_settings';
