// @datocms/cma-client/dist/cjs/generated/resources/SubscriptionLimit.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Utils = __importStar(require("@datocms/rest-client-utils"));
const BaseResource_1 = __importDefault(require("../../BaseResource"));
class SubscriptionLimit extends BaseResource_1.default {
    /**
     * Get all the subscription limits
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/subscription-limit/instances
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    list() {
        return this.rawList().then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Get all the subscription limits
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/subscription-limit/instances
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawList() {
        return this.client.request({
            method: 'GET',
            url: '/subscription-limits',
        });
    }
    /**
     * Get a single subscription limit
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/subscription-limit/self
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    find(subscriptionLimitId) {
        return this.rawFind(Utils.toId(subscriptionLimitId)).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Get a single subscription limit
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/subscription-limit/self
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawFind(subscriptionLimitId) {
        return this.client.request({
            method: 'GET',
            url: `/subscription-limits/${subscriptionLimitId}`,
        });
    }
}
exports.default = SubscriptionLimit;
SubscriptionLimit.TYPE = 'subscription_limit';
