// @datocms/cma-client/dist/cjs/generated/resources/UploadCollection.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Utils = __importStar(require("@datocms/rest-client-utils"));
const BaseResource_1 = __importDefault(require("../../BaseResource"));
class UploadCollection extends BaseResource_1.default {
    /**
     * Create a new upload collection
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/upload-collection/create
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    create(body) {
        return this.rawCreate(Utils.serializeRequestBody(body, {
            type: 'upload_collection',
            attributes: ['label', 'position'],
            relationships: ['parent'],
        })).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Create a new upload collection
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/upload-collection/create
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawCreate(body) {
        return this.client.request({
            method: 'POST',
            url: '/upload-collections',
            body,
        });
    }
    /**
     * Update a upload collection
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/upload-collection/update
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    update(uploadCollectionId, body) {
        return this.rawUpdate(Utils.toId(uploadCollectionId), Utils.serializeRequestBody(body, {
            id: Utils.toId(uploadCollectionId),
            type: 'upload_collection',
            attributes: ['label', 'position'],
            relationships: ['parent', 'children'],
        })).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Update a upload collection
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/upload-collection/update
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawUpdate(uploadCollectionId, body) {
        return this.client.request({
            method: 'PUT',
            url: `/upload-collections/${uploadCollectionId}`,
            body,
        });
    }
    /**
     * List all upload collections
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/upload-collection/instances
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    list(queryParams) {
        return this.rawList(queryParams).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * List all upload collections
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/upload-collection/instances
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawList(queryParams) {
        return this.client.request({
            method: 'GET',
            url: '/upload-collections',
            queryParams,
        });
    }
    /**
     * Retrieve a upload collection
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/upload-collection/self
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    find(uploadCollectionId) {
        return this.rawFind(Utils.toId(uploadCollectionId)).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Retrieve a upload collection
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/upload-collection/self
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawFind(uploadCollectionId) {
        return this.client.request({
            method: 'GET',
            url: `/upload-collections/${uploadCollectionId}`,
        });
    }
    /**
     * Delete a upload collection
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/upload-collection/destroy
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    destroy(uploadCollectionId) {
        return this.rawDestroy(Utils.toId(uploadCollectionId)).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Delete a upload collection
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/upload-collection/destroy
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawDestroy(uploadCollectionId) {
        return this.client.request({
            method: 'DELETE',
            url: `/upload-collections/${uploadCollectionId}`,
        });
    }
    /**
     * Reorders a set of upload collections
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/upload-collection/reorder
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     *
     * @deprecated This API call is to be considered private and might change without notice
     */
    reorder(body) {
        return this.rawReorder(Utils.serializeRequestBody(body, {
            type: 'upload_collection',
            attributes: ['position'],
            relationships: ['parent'],
        })).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Reorders a set of upload collections
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/upload-collection/reorder
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     *
     * @deprecated This API call is to be considered private and might change without notice
     */
    rawReorder(body) {
        return this.client.request({
            method: 'POST',
            url: '/upload-collections/reorder',
            body,
        });
    }
}
exports.default = UploadCollection;
UploadCollection.TYPE = 'upload_collection';
