// @datocms/cma-client-node/dist/cjs/utils/uploadLocalFileToS3.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.uploadLocalFileToS3 = void 0;
const got_1 = __importStar(require("got"));
const fs_1 = require("fs");
const mime_types_1 = __importDefault(require("mime-types"));
const rest_client_utils_1 = require("@datocms/rest-client-utils");
const rest_client_utils_2 = require("@datocms/rest-client-utils");
function uploadLocalFileToS3(filePath, url, { onProgress } = {}) {
    let isCanceled = false;
    let putPromise;
    return (0, rest_client_utils_2.makeCancelablePromise)(() => __awaiter(this, void 0, void 0, function* () {
        if (isCanceled) {
            throw new rest_client_utils_1.CanceledPromiseError();
        }
        const { size: totalLength } = yield fs_1.promises.stat(filePath);
        if (isCanceled) {
            throw new rest_client_utils_1.CanceledPromiseError();
        }
        try {
            putPromise = got_1.default.put(url, {
                headers: {
                    'Content-Type': mime_types_1.default.lookup(filePath) || 'application/octet-stream',
                    'Content-Length': `${totalLength}`,
                },
                responseType: 'json',
                body: (0, fs_1.createReadStream)(filePath),
            });
        }
        catch (e) {
            if (e instanceof got_1.CancelError) {
                throw new rest_client_utils_1.CanceledPromiseError();
            }
            else {
                throw e;
            }
        }
        if (onProgress) {
            putPromise.on('uploadProgress', ({ percent }) => {
                if (!isCanceled) {
                    onProgress({
                        type: 'UPLOADING_FILE',
                        payload: { progress: Math.round(percent * 100) },
                    });
                }
            });
        }
        yield putPromise;
    }), () => {
        isCanceled = true;
        if (putPromise) {
            putPromise.cancel();
        }
    });
}
exports.uploadLocalFileToS3 = uploadLocalFileToS3;
