// @datocms/cma-client-node/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CanceledPromiseError = exports.BaseClient = exports.LogLevel = exports.ApiError = exports.buildBaseClient = exports.buildBlockRecord = exports.Resources = exports.SimpleSchemaTypes = exports.SchemaTypes = exports.downloadFile = exports.uploadLocalFileAndReturnPath = void 0;
__exportStar(require("./Client"), exports);
__exportStar(require("./buildClient"), exports);
__exportStar(require("./utils/uploadLocalFileAndReturnPath"), exports);
__exportStar(require("./UploadResource"), exports);
__exportStar(require("./utils/uploadLocalFileToS3"), exports);
var uploadLocalFileAndReturnPath_1 = require("./utils/uploadLocalFileAndReturnPath");
Object.defineProperty(exports, "uploadLocalFileAndReturnPath", { enumerable: true, get: function () { return uploadLocalFileAndReturnPath_1.uploadLocalFileAndReturnPath; } });
var downloadFile_1 = require("./utils/downloadFile");
Object.defineProperty(exports, "downloadFile", { enumerable: true, get: function () { return downloadFile_1.downloadFile; } });
var cma_client_1 = require("@datocms/cma-client");
Object.defineProperty(exports, "SchemaTypes", { enumerable: true, get: function () { return cma_client_1.SchemaTypes; } });
Object.defineProperty(exports, "SimpleSchemaTypes", { enumerable: true, get: function () { return cma_client_1.SimpleSchemaTypes; } });
Object.defineProperty(exports, "Resources", { enumerable: true, get: function () { return cma_client_1.Resources; } });
Object.defineProperty(exports, "buildBlockRecord", { enumerable: true, get: function () { return cma_client_1.buildBlockRecord; } });
Object.defineProperty(exports, "buildBaseClient", { enumerable: true, get: function () { return cma_client_1.buildClient; } });
Object.defineProperty(exports, "ApiError", { enumerable: true, get: function () { return cma_client_1.ApiError; } });
Object.defineProperty(exports, "LogLevel", { enumerable: true, get: function () { return cma_client_1.LogLevel; } });
Object.defineProperty(exports, "BaseClient", { enumerable: true, get: function () { return cma_client_1.Client; } });
var rest_client_utils_1 = require("@datocms/rest-client-utils");
Object.defineProperty(exports, "CanceledPromiseError", { enumerable: true, get: function () { return rest_client_utils_1.CanceledPromiseError; } });
