// @datocms/cma-client/dist/esm/generated/resources/AuditLogEvent.js
import * as Utils from '@datocms/rest-client-utils';
import BaseResource from '../../BaseResource';
export default class AuditLogEvent extends BaseResource {
    /**
     * List Audit Log events
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/audit-log_event/query
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    query(body) {
        return this.rawQuery(Utils.serializeRequestBody(body, {
            type: 'audit_log_query',
            attributes: ['filter', 'next_token', 'detailed_log'],
            relationships: [],
        })).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * List Audit Log events
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/audit-log_event/query
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawQuery(body) {
        return this.client.request({
            method: 'POST',
            url: '/audit-log-events/query',
            body,
        });
    }
}
AuditLogEvent.TYPE = 'audit_log_event';
