// @datocms/cma-client/dist/esm/generated/resources/ItemVersion.js
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var __await = (this && this.__await) || function (v) { return this instanceof __await ? (this.v = v, this) : new __await(v); }
var __asyncGenerator = (this && this.__asyncGenerator) || function (thisArg, _arguments, generator) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var g = generator.apply(thisArg, _arguments || []), i, q = [];
    return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
    function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
    function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
    function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
    function fulfill(value) { resume("next", value); }
    function reject(value) { resume("throw", value); }
    function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
};
import * as Utils from '@datocms/rest-client-utils';
import BaseResource from '../../BaseResource';
export default class ItemVersion extends BaseResource {
    /**
     * Restore an old record version
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/item-version/restore
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    restore(itemVersionId) {
        return this.rawRestore(Utils.toId(itemVersionId)).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Restore an old record version
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/item-version/restore
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawRestore(itemVersionId) {
        return this.client.request({
            method: 'POST',
            url: `/versions/${itemVersionId}/restore`,
        });
    }
    /**
     * List all record versions
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/item-version/instances
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    list(itemId, queryParams) {
        return this.rawList(Utils.toId(itemId), queryParams).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * List all record versions
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/item-version/instances
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawList(itemId, queryParams) {
        return this.client.request({
            method: 'GET',
            url: `/items/${itemId}/versions`,
            queryParams,
        });
    }
    /**
     * Async iterator to auto-paginate over elements returned by list()
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/item-version/instances
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    listPagedIterator(itemId, queryParams, iteratorOptions) {
        return __asyncGenerator(this, arguments, function* listPagedIterator_1() {
            var _a, e_1, _b, _c;
            try {
                for (var _d = true, _e = __asyncValues(this.rawListPagedIterator(Utils.toId(itemId), queryParams, iteratorOptions)), _f; _f = yield __await(_e.next()), _a = _f.done, !_a;) {
                    _c = _f.value;
                    _d = false;
                    try {
                        const element = _c;
                        yield yield __await(Utils.deserializeJsonEntity(element));
                    }
                    finally {
                        _d = true;
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (!_d && !_a && (_b = _e.return)) yield __await(_b.call(_e));
                }
                finally { if (e_1) throw e_1.error; }
            }
        });
    }
    /**
     * Async iterator to auto-paginate over elements returned by rawList()
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/item-version/instances
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawListPagedIterator(itemId, queryParams, iteratorOptions) {
        Utils.warnOnPageQueryParam(queryParams);
        return Utils.rawPageIterator({
            defaultLimit: 15,
            maxLimit: 50,
        }, (page) => this.rawList(itemId, Object.assign(Object.assign({}, queryParams), { page })), iteratorOptions);
    }
    /**
     * Retrieve a record version
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/item-version/self
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    find(itemVersionId, queryParams) {
        return this.rawFind(Utils.toId(itemVersionId), queryParams).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Retrieve a record version
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/item-version/self
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawFind(itemVersionId, queryParams) {
        return this.client.request({
            method: 'GET',
            url: `/versions/${itemVersionId}`,
            queryParams,
        });
    }
}
ItemVersion.TYPE = 'item_version';
