// @datocms/cma-client/dist/esm/generated/resources/MaintenanceMode.js
import * as Utils from '@datocms/rest-client-utils';
import BaseResource from '../../BaseResource';
export default class MaintenanceMode extends BaseResource {
    /**
     * Retrieve maintenence mode
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/maintenance-mode/self
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    find() {
        return this.rawFind().then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Retrieve maintenence mode
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/maintenance-mode/self
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawFind() {
        return this.client.request({
            method: 'GET',
            url: '/maintenance-mode',
        });
    }
    /**
     * Activate maintenance mode: this means that the primary environment will be read-only
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/maintenance-mode/activate
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    activate(queryParams) {
        return this.rawActivate(queryParams).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Activate maintenance mode: this means that the primary environment will be read-only
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/maintenance-mode/activate
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawActivate(queryParams) {
        return this.client.request({
            method: 'PUT',
            url: '/maintenance-mode/activate',
            queryParams,
        });
    }
    /**
     * De-activate maintenance mode
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/maintenance-mode/deactivate
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    deactivate() {
        return this.rawDeactivate().then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * De-activate maintenance mode
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/maintenance-mode/deactivate
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawDeactivate() {
        return this.client.request({
            method: 'PUT',
            url: '/maintenance-mode/deactivate',
        });
    }
}
MaintenanceMode.TYPE = 'maintenance_mode';
