// @datocms/cma-client/dist/esm/generated/resources/Session.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import * as Utils from '@datocms/rest-client-utils';
import BaseResource from '../../BaseResource';
var Session = /** @class */ (function (_super) {
    __extends(Session, _super);
    function Session() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Create a new session
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/session/create
     *
     * @deprecated This API call is to be considered private and might change without notice
     */
    Session.prototype.create = function (body) {
        return this.rawCreate(Utils.serializeRequestBody(body, {
            type: 'email_credentials',
            attributes: ['email', 'password', 'otp_code'],
            relationships: [],
        })).then(function (body) {
            return Utils.deserializeResponseBody(body);
        });
    };
    /**
     * Create a new session
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/session/create
     *
     * @deprecated This API call is to be considered private and might change without notice
     */
    Session.prototype.rawCreate = function (body) {
        return this.client.request({
            method: 'POST',
            url: "/sessions",
            body: body,
        });
    };
    Session.TYPE = 'session';
    return Session;
}(BaseResource));
export default Session;
