// @datocms/cma-client/dist/esm/generated/resources/SubscriptionLimit.js
import * as Utils from '@datocms/rest-client-utils';
import BaseResource from '../../BaseResource';
export default class SubscriptionLimit extends BaseResource {
    /**
     * Get all the subscription limits
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/subscription-limit/instances
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    list() {
        return this.rawList().then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Get all the subscription limits
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/subscription-limit/instances
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawList() {
        return this.client.request({
            method: 'GET',
            url: '/subscription-limits',
        });
    }
    /**
     * Get a single subscription limit
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/subscription-limit/self
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    find(subscriptionLimitId) {
        return this.rawFind(Utils.toId(subscriptionLimitId)).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Get a single subscription limit
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/subscription-limit/self
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawFind(subscriptionLimitId) {
        return this.client.request({
            method: 'GET',
            url: `/subscription-limits/${subscriptionLimitId}`,
        });
    }
}
SubscriptionLimit.TYPE = 'subscription_limit';
