// @datocms/cma-client/dist/esm/generated/resources/UploadRequest.js
import * as Utils from '@datocms/rest-client-utils';
import BaseResource from '../../BaseResource';
export default class UploadRequest extends BaseResource {
    /**
     * Request a new permission to upload a file
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/upload-request/create
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    create(body) {
        return this.rawCreate(Utils.serializeRequestBody(body, {
            type: 'upload_request',
            attributes: ['filename'],
            relationships: ['upload_collection'],
        })).then((body) => Utils.deserializeResponseBody(body));
    }
    /**
     * Request a new permission to upload a file
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/upload-request/create
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    rawCreate(body) {
        return this.client.request({
            method: 'POST',
            url: '/upload-requests',
            body,
        });
    }
}
UploadRequest.TYPE = 'upload_request';
