// @datocms/cma-client-node/dist/cjs/generated/resources.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Workflow = exports.Site = exports.UploadSmartTag = exports.UploadTag = exports.UsageCounter = exports.DailyUsage = exports.PublicInfo = exports.WhiteLabelSettings = exports.EmojiSuggestions = exports.SsoSettings = exports.SsoGroup = exports.EditingSession = exports.SiteInvitation = exports.UploadFilter = exports.ItemTypeFilter = exports.SearchIndex = exports.BuildTrigger = exports.WebhookCall = exports.Webhook = exports.MaintenanceMode = exports.Environment = exports.SearchResult = exports.ScheduledUnpublishing = exports.ScheduledPublication = exports.UploadTrack = exports.UploadRequest = exports.Upload = exports.ItemVersion = exports.Item = exports.SearchIndexEvent = exports.BuildEvent = exports.SubscriptionFeature = exports.SubscriptionLimit = exports.JobResult = exports.Plugin = exports.AccessToken = exports.Session = exports.Fieldset = exports.Field = exports.ItemType = exports.UploadCollection = exports.SchemaMenuItem = exports.MenuItem = exports.AuditLogEvent = exports.SsoUser = exports.User = exports.Role = void 0;
const cma_client_1 = require("@datocms/cma-client");
exports.Role = cma_client_1.Resources.Role;
exports.User = cma_client_1.Resources.User;
exports.SsoUser = cma_client_1.Resources.SsoUser;
exports.AuditLogEvent = cma_client_1.Resources.AuditLogEvent;
exports.MenuItem = cma_client_1.Resources.MenuItem;
exports.SchemaMenuItem = cma_client_1.Resources.SchemaMenuItem;
exports.UploadCollection = cma_client_1.Resources.UploadCollection;
exports.ItemType = cma_client_1.Resources.ItemType;
exports.Field = cma_client_1.Resources.Field;
exports.Fieldset = cma_client_1.Resources.Fieldset;
exports.Session = cma_client_1.Resources.Session;
exports.AccessToken = cma_client_1.Resources.AccessToken;
exports.Plugin = cma_client_1.Resources.Plugin;
exports.JobResult = cma_client_1.Resources.JobResult;
exports.SubscriptionLimit = cma_client_1.Resources.SubscriptionLimit;
exports.SubscriptionFeature = cma_client_1.Resources.SubscriptionFeature;
exports.BuildEvent = cma_client_1.Resources.BuildEvent;
exports.SearchIndexEvent = cma_client_1.Resources.SearchIndexEvent;
exports.Item = cma_client_1.Resources.Item;
exports.ItemVersion = cma_client_1.Resources.ItemVersion;
var Upload_1 = require("../resources/Upload");
Object.defineProperty(exports, "Upload", { enumerable: true, get: function () { return __importDefault(Upload_1).default; } });
exports.UploadRequest = cma_client_1.Resources.UploadRequest;
var UploadTrack_1 = require("../resources/UploadTrack");
Object.defineProperty(exports, "UploadTrack", { enumerable: true, get: function () { return __importDefault(UploadTrack_1).default; } });
exports.ScheduledPublication = cma_client_1.Resources.ScheduledPublication;
exports.ScheduledUnpublishing = cma_client_1.Resources.ScheduledUnpublishing;
exports.SearchResult = cma_client_1.Resources.SearchResult;
exports.Environment = cma_client_1.Resources.Environment;
exports.MaintenanceMode = cma_client_1.Resources.MaintenanceMode;
exports.Webhook = cma_client_1.Resources.Webhook;
exports.WebhookCall = cma_client_1.Resources.WebhookCall;
exports.BuildTrigger = cma_client_1.Resources.BuildTrigger;
exports.SearchIndex = cma_client_1.Resources.SearchIndex;
exports.ItemTypeFilter = cma_client_1.Resources.ItemTypeFilter;
exports.UploadFilter = cma_client_1.Resources.UploadFilter;
exports.SiteInvitation = cma_client_1.Resources.SiteInvitation;
exports.EditingSession = cma_client_1.Resources.EditingSession;
exports.SsoGroup = cma_client_1.Resources.SsoGroup;
exports.SsoSettings = cma_client_1.Resources.SsoSettings;
exports.EmojiSuggestions = cma_client_1.Resources.EmojiSuggestions;
exports.WhiteLabelSettings = cma_client_1.Resources.WhiteLabelSettings;
exports.PublicInfo = cma_client_1.Resources.PublicInfo;
exports.DailyUsage = cma_client_1.Resources.DailyUsage;
exports.UsageCounter = cma_client_1.Resources.UsageCounter;
exports.UploadTag = cma_client_1.Resources.UploadTag;
exports.UploadSmartTag = cma_client_1.Resources.UploadSmartTag;
exports.Site = cma_client_1.Resources.Site;
exports.Workflow = cma_client_1.Resources.Workflow;
